Fixes #28078 - add event subscribers api

kamils-iRonin opened the pull request:
This PR is continuation of #7112  

TODO: 
- [x] possibility to fire a hook that is unrelated to the model lifecycle

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

timogoebel committed:
fixes #28078 - add event subscribers api

lzap commented on test/models/concerns/foreman/hook_subscribers/observable_test.rb:
```diff
@@ -0,0 +1,111 @@
+    let(:model) { klass.create(name: 'My Model') }
+
+    setup do
+      Foreman::Plugin.event_observers_registry.register_event_observer('ObservableTest::DummyObserver')
```
Why `ObservableTest::DummyObserver` is a string and not constant? I must be missing something.

lzap commented on app/models/concerns/foreman/hook_subscribers/observable.rb:
```diff
@@ -0,0 +1,47 @@
+        end
+
+        def event_name_for(hook_name)
+          "#{name.underscore}/#{hook_name}"
```
Is there any intention to use `/` characters for some tree-like structure? Honestly, I'd discourage you from doing that, if there are no objections let's make events named via underscore: `subnet_created` or `subnet_deleted` to have more of a flat-like naming scheme.

kamils-iRonin replied:
We use `name.underscore` as the event name prefix, so for namespaced classes, e.g. `Subnet::Ipv4` it is `subnet/ipv4`. When I think about it now, the name of the event should not depend on the name of the class from which the event is sent. The easiest way is to send an event named just like the hook name, eg.
```ruby
set_hook :model_updated, on: :update
```
should send an event named `model_updated`, and
```ruby
model.trigger_hook(:hook_triggered)
```
should send an event named `hook_triggered`.

lzap commented on app/models/concerns/foreman/hook_subscribers/observable.rb:
```diff
@@ -0,0 +1,47 @@
+      private
+
+      def payload_for(payload, blk)
+        payload || blk&.call(self) || { id: id }
```
I am still in doubt if it's good idea to pass whole instance as a payload for ActiveRecord callbacks. Previously, some model changes caused unexpected regressions in user's hooks. Programmers tend to forget that a small change in model that is handled correctly in all controllers/views can cause this.

On the other hand, sending only ID is not very practical. I am not sure there's a "good" solution.

lzap commented on app/models/concerns/foreman/hook_subscribers/observable.rb:
```diff
@@ -0,0 +1,47 @@
+      end
+    end
+  end
+end
```
I think Rails Instrumentation API would be the best fit, on the other hand this allows full control of the registration process and it's also good approach.

lzap commented on test/models/concerns/foreman/hook_subscribers/observable_test.rb:
```diff
@@ -0,0 +1,111 @@
+        set_hook :updated_custom_payload, on: :update, payload: { custom_payload: true }
+        set_hook :updated_block_payload, on: :update do |m|
+          { id: m.id, block: true }
+        end
```
Wow, the syntax of hook 2 and 3 is really close. Okay :-)

lzap commented on app/models/model.rb:
```diff
@@ -4,6 +4,13 @@ class Model < ApplicationRecord
+  end
 
   before_destroy EnsureNotUsedBy.new(:hosts)
   has_many_hosts
```
One think is missing tho - we would like to have subsciption API for AR models as well as regular Ruby objects. The Observable class only works with AR (`after_commit`), please either refactor it or create a different Observable class for regular Ruby objects. Basically, I'd expect that it would define some "triggering" methods which could be called programmatically. Also add a test for it.

This is where I think `model/action` naming convention is weird, I think a hook should be simply a string that's ensured to be just `[a-z0-9_]` for example `model_created` so implementations can take advantage of it - e.g. "shell" plugin would expect a filename named like this.

lzap commented on test/models/concerns/foreman/hook_subscribers/observable_test.rb:
```diff
@@ -0,0 +1,111 @@
+
+      expected_event = { event_name: 'model/updated_default_payload', payload: { id: model.id }}
+
+      DummyObserver.any_instance.expects(:notify).with(expected_event)
```
Why exactly you stub the `notify` method first and then change your mine to `expect` it? I usually do this in `setup` method, that's fine. But in every test it looks weird to me.

lzap reviewed (changes requested):
I think it's almost there.

lzap commented:
One more thing - we need to properly define what the expected payloads are for our ActiveRecord hooks. Tests don't test it, they use dummy object. Could you create an extra test:

* create a dummy (but valid) host
* make an action
* compare payload to expected structure

Now, we haven't defined payload for AR models yet. Old hooks plugin leveraged API controller RABL templates to render its contents in JSON, however this created problems. There are several options:

* serialize whole model with no associations into JSON
* serialize whole model and loaded associations into JSON (I think this is wrong)
* create explicit methods `to_hook_payload` and call it (models without the method would only return ID).
* keep using RABL (take care, it bites!)
* other ideas

@timogoebel

timogoebel commented:
> we need to properly define what the expected payloads are for our ActiveRecord hooks

When we discussed this last time (🍻, cheers!) we thought that for creating/update events we would just publish the object's id so the client can query the required data in a separate API call. We quickly noticed that this does not work for deletion events and came up with the idea that a developer should define the payload (as a block) explicitly when defining the hook.

kamils-iRonin committed:
fixes #28078 - introduce Rails Instrumentation

kamils-iRonin commented:
@lzap Can you please take a look? I've refactored the code to use the Rails Instrumentation. It is also possible to trigger a hook on a regular Ruby object. Also, the name of the event no longer depends on the name of the class from which it is sent. I have introduced the `subscribers` folder, this can be a good place if we want to use subscribers in the core.

kamils-iRonin commented on lib/foreman/observable.rb:
```diff
@@ -0,0 +1,19 @@
+    end
+
+    def do_notify_event_observers(event_name, payload:)
+      ActiveSupport::Notifications.instrument(event_name, payload)
```
I think it would make sense to add `foreman` namespace so we can filter only events sent from our API by using `/^foreman./` regex
```suggestion
      ActiveSupport::Notifications.instrument("foreman.#{event_name}", payload)
```

lzap replied:
Yes, please do.

kamils-iRonin replied:
actually, to follow the rails naming convention which is `event.library` we should add `foreman` namespace as a suffix

kamils-iRonin commented:
[test foreman]

ares commented on developer_docs/how_to_create_a_plugin.asciidoc:
```diff
@@ -2092,6 +2092,20 @@ Similarly for mutations:
+[[adding-subscribers]]
+=== Adding subscribers
+
+_Requires Foreman 1.25 or higher, set requires_foreman '>= 1.25' in
```
this may now be 2.0

lzap replied:
I'd just delete this line, it's dev doc anyway.

ares commented on developer_docs/how_to_create_a_plugin.asciidoc:
```diff
@@ -2092,6 +2092,20 @@ Similarly for mutations:
+  subscribe :my_event, MyPlugin::MySubscriber
+....
+
+where `:my_event` is the name of the event you want to subscribe to. You may also subscribe to multiple events at once by using a regular expression.
```
could you please add also example for this?

ares commented on developer_docs/how_to_create_a_plugin.asciidoc:
```diff
@@ -2092,6 +2092,20 @@ Similarly for mutations:
+
+[source, ruby]
+....
+  subscribe :my_event, MyPlugin::MySubscriber
```
a bit more context where I put this line would be good, also perhaps mentioning how can I find all existing hooks in core, something like grep for "set_hook"

ares reviewed (commented):
Big thanks for adding developer documentation, few suggestions left as comments. Also one question, is adding hooks for HW model just an example or something that will be used already somewhere?

kamils-iRonin commented:
@ares Thank you for your feedback. I've updated documentation as you suggested.

> is adding hooks for HW model just an example or something that will be used already somewhere?

it's just an example

lzap commented:
Apologies for the delay, I am going to do last read and test this.

lzap commented on developer_docs/how_to_create_a_plugin.asciidoc:
```diff
@@ -2092,6 +2092,30 @@ Similarly for mutations:
+  subscribe /^foreman./, MyPlugin::MySubscriber
+....
+
+You can find all observable events by calling `Foreman::EventSubscribers.all_observable_events` in the Rails console. Keep in mind that this list only contains events that were set by the `set_hook` method.
```
This is giving me without the "foreman." prefix, so I guess it's wrong. Need to test this.

lzap reviewed (changes requested):
Looks great. One last thing - we only add model events with this patch. I think it is worth adding few custom ones:

* When a host enters or exits build mode (two events named `build_entered` and `build_exited`, host id and hostname (FQDN) could be both part of the payload for this one)
* When our orchestration task finishes execution (e.g. DHCP record added)
* When a host status is changed
* When a user reaches limit of maximum password failures and is blocked

These are just some ideas, you don't need to implement all I think the most useful is the first one as this is constantly a pain for users to hook this via active record.

I am going to write a plugin now to replace foreman_hooks.

kamils-iRonin commented:
@lzap I added `build_entered` and `build_exited` hooks. I think we should add more hooks in a separate PR to not make this one too complicated. I've also refactored the code a bit

lzap commented:
Absolutely, thanks. Writing my implementation.

lzap commented:
Do you have a plugin example git repo (e.g. webhooks which implements this changed API) by chance? So I could copy and paste some files :-)

lzap commented on app/subscribers/base_subscriber.rb:
```diff
@@ -0,0 +1,11 @@
+  def call(event)
+    raise NotImplementedError
+  end
+end
```
Nitpick: I wonder if we should move this to Foreman namespace, chances are Rails might introduce this name in the future.

kamils-iRonin commented:
> Do you have a plugin example git repo (e.g. webhooks which implements this changed API) by chance? So I could copy and paste some files :-)

I have opened the PR to `foreman_webhooks` with necessary changes https://github.com/timogoebel/foreman_webhooks/pull/6

lzap commented:
I must be doing something wrong. Implemented a catch-them-all pokémon-style subscriber:

```
    initializer 'foreman_redishooks.register_plugin', before: :finisher_hook do |_app|
      Foreman::Plugin.register :foreman_redishooks do
        requires_foreman '>= 1.24'

        subscribe(/.foreman$/, ForemanRedishooks::RedisSubscriber)
      end
    end
```

But during host creation I only see build event:

CATCH #<ActiveSupport::Notifications::Event:0x000055555aaf6bf8 @name="build_entered.foreman", @payload={:id=>6, :hostname=>"simple-5445.nat.lan"}, @time=2020-01-13 16:19:33 +0100, @transaction_id="dc4b020d343dc595c882", @end=2020-01-13 16:19:33 +0100, @children=[], @duration=nil>

lzap commented on test/models/hosts/managed_test.rb:
```diff
@@ -122,5 +122,31 @@ class ManagedTest < ActiveSupport::TestCase
+      end
+    end
   end
 end
```
Can you also add an actual AR event test? Looks like it might not get sent actually.

kamils-iRonin replied:
I'm not sure how it is supposed to work and how to test it. What do you think about this:
```ruby
describe 'AR hooks' do
  let(:host) { FactoryBot.build(:host, :managed) }

  test 'call #build_hooks on commit' do
    host.expects(:build_hooks)

    host.run_callbacks(:commit)
  end

  describe '#build_hooks' do
    context 'when build mode is enabled' do
      it 'runs :build callbacks' do
        host.stubs(:build?).returns(true)
        host.stubs(:previous_changes).returns({ 'build' => [false, true] })

        host.expects(:run_callbacks).with(:build)

        host.build_hooks
      end
    end

    context 'when build mode is disabled' do
      it 'runs :provision callbacks' do
        host.stubs(:build?).returns(false)
        host.stubs(:previous_changes).returns({ 'build' => [true, false] })

        host.expects(:run_callbacks).with(:provision)

        host.build_hooks
      end
    end
  end
end
```
Do you think it makes any sense?

lzap replied:
I thought literally creating a record in the DB, this mocking is not very useful. Some Rails AR regressions hit us in the past but you know what let's just ignore this we would easily find out.

lzap commented on config/initializers/subscribers.rb:
```diff
@@ -0,0 +1,5 @@
+
+ActiveSupport::Notifications.tap do |config|
+  # config.subscribe :my_event, MySubscriber
+end
```
Is this a leftover? I'd encourage engs to write plugins if they want to subscribe. Core should only provide minimum contract I think.

lzap commented:
After a short chat with Timo, it looks like we agreed to opt-in approach for models. Can you add hosts and hostgroups events? That should get us started, other models we can add later on.

lzap commented:
One more thing to consider, I guess the transaction_id="dc4b020d343dc595c882" is some Rails Subscriber API internal, can we also send request_id and session_id which we have stored in Logging.mdc context? I think we could actually send all of the context variables along (organization, location). This could be quite useful when debugging stuff to correlate messages and transactions in logs.

This would be part of the payload by default.

lzap commented on developer_docs/how_to_create_a_plugin.asciidoc:
```diff
@@ -2092,6 +2092,30 @@ Similarly for mutations:
+
+[source, ruby]
+....
+  subscribe /.foreman$/, MyPlugin::MySubscriber
```
Is there any particular reason why this is suffix and not prefix? It's non-standard, I'd not expect that. I am not sure if Redis even supports that (all examples shows "TEST*" wildcard notation I need to test it).

Also I'd appreciate if this also included "event" because we might add more using Rails Subscribtion API later and this really captures the whole "namespace". So ideally: "foreman.event.NAME" if you agree with that.

Edit: It seems to support it however it feels weird to me: https://redis.io/commands/psubscribe

kamils-iRonin replied:
It's also a bit weird to me, but I added the namespace as suffix to follow the Rails naming convention, which is `event.library`. [[1]](https://guides.rubyonrails.org/active_support_instrumentation.html#creating-custom-events) There are two ways to subscribe to events:

- `ActiveSupport::Notifications.subscribe 'my_event.my_namespace'`, which we use currently and which I think gives more possibilities, I like that we can use regex to subscribe to multiple events at once. 
- a class that inherits from `ActiveSupport::Subscriber`, e.g.

```ruby
class MySubscriber < ActiveSupport::Subscriber
  attach_to :my_namespace

  def my_event(event)
    puts event.inspect
  end
end
```
You must attach a subscriber to the namespace and create a method named like the event you want to subscribe to. The `my_event` method will be executed whenever the `my_event.my_namespace` event is published. Note that the namespace needs to be added as suffix in this case.

Of course, as we use regex, nothing can stop us from adding namespace as a prefix. So it is possible to name events like you suggested, `foreman.event.NAME`. WDYT?

lzap replied:
If you followed Rails conventions than it's probably fine although not all things in Rails are great :-)

timogoebel reviewed (dismissed):
Looks good to me.
@lzap: Is this ready?

lzap commented on lib/foreman/observable.rb:
```diff
@@ -0,0 +1,24 @@
+module Foreman
+  module Observable
+    DEFAULT_NAMESPACE = :foreman
```
Last minor thing can you make this "event.foreman" or "foreman.event"? Just in case we would use the API for something else in Foreman, this seems to be pretty generic.

lzap reviewed (dismissed):
Apologies, I haven't been able to finish my Redis-based implementation but let's merge this this is ready for core and I will finish it later.

lzap reviewed (approved):
It works fine, merging. Thanks for the patience with me. We only have host/hostgroups implemented but that's a good start.

lzap commented:
A short followup after a meeting we had with @ehelms and the group: to publish an arbitrary event the plan was to use simple approach just to call `trigger_hook` directly. However there is no registry of those "custom" hooks, there should probably be one so a plugin can easily list all available hooks.

The plan is to get devs playing with this early because foreman_hooks is one of the most popular plugins and we would like to get more feedback on this. I am currently writing a post about this one.

ehelms commented:
Thanks for capturing this as it hits right at the point I wanted to make. This doesn't seem like it's a lot of work, but I wanted to ensure this covered all use cases from the plugin API perspective

timogoebel commented:
@ehelms, @lzap : There is a all_observable_events method to list all events. Is there more you‘d like to have?
